lib/Controller/ConsentController.php

```diff
@@ -0,0 +1,358 @@
+            }
+            Stats::log('consent:accept', $statsInfo);
+
+            if (
+                array_key_exists('consent:store', $state)
+                && $request->query->get('saveconsent') === '1'
+            ) {
+                // Save consent
+                $store = $state['consent:store'];
+                $userId = $state['consent:store.userId'];
+                $targetedId = $state['consent:store.destination'];
+                $attributeSet = $state['consent:store.attributeSet'];
+
+                $this->logger::debug(
+                    'Consent - saveConsent() : [' . $userId . '|' . $targetedId . '|' . $attributeSet . ']'
+                );
+                try {
+                    $store->saveConsent($userId, $targetedId, $attributeSet);
+                } catch (Exception $e) {
+                    $this->logger::error('Consent: Error writing to storage: ' . $e->getMessage());
+                }
+            }
+
+            return new RunnableResponse([Auth\ProcessingChain::class, 'resumeProcessing'], [$state]);
+        }
+
+        // Prepare attributes for presentation
+        $attributes = $state['Attributes'];
+        $noconsentattributes = $state['consent:noconsentattributes'];
+
+        // Remove attributes that do not require consent
+        foreach ($attributes as $attrkey => $attrval) {
+            if (in_array($attrkey, $noconsentattributes, true)) {
+                unset($attributes[$attrkey]);
+            }
+        }
+        $para = [
+            'attributes' => &$attributes
+        ];
+
+        // Reorder attributes according to attributepresentation hooks
+        Module::callHooks('attributepresentation', $para);
+
+        // Parse parameters
+        if (array_key_exists('name', $state['Source'])) {
+            $srcName = $state['Source']['name'];
+        } elseif (array_key_exists('OrganizationDisplayName', $state['Source'])) {
+            $srcName = $state['Source']['OrganizationDisplayName'];
+        } else {
+            $srcName = $state['Source']['entityid'];
+        }
+
+        if (array_key_exists('name', $state['Destination'])) {
+            $dstName = $state['Destination']['name'];
+        } elseif (array_key_exists('OrganizationDisplayName', $state['Destination'])) {
+            $dstName = $state['Destination']['OrganizationDisplayName'];
+        } else {
+            $dstName = $state['Destination']['entityid'];
+        }
+
+        // Make, populate and layout consent form
+        $t = new Template($this->config, 'consent:consentform.twig');
+        $translator = $t->getTranslator();
+        $t->data['srcMetadata'] = $state['Source'];
+        $t->data['dstMetadata'] = $state['Destination'];
+        $t->data['yesTarget'] = Module::getModuleURL('consent/getconsent');
+        $t->data['yesData'] = ['StateId' => $stateId];
+        $t->data['noTarget'] = Module::getModuleURL('consent/noconsent');
+        $t->data['noData'] = ['StateId' => $stateId];
```

Comment:
A lot of these are not used in the template I think... duplications like yesdata/nodata etc